src/main/java/org/openjdk/jextract/impl/CDeclarationPrinter.java

```diff
@@ -0,0 +1,261 @@
+    static String declaration(Declaration decl, String prefix) {
+        Objects.requireNonNull(decl);
+        Objects.requireNonNull(prefix);
+        return new CDeclarationPrinter(prefix).print(decl);
+    }
+
+    static String declaration(Type.Function funcType, String name) {
+        return nameAndType(funcType, "*" + name);
+    }
+
+    @Override
+    public Void visitScoped(Declaration.Scoped d, Void ignored) {
+        indent();
+        var tag = typeTag(d);
+        if (!tag.isEmpty()) {
+            builder.append(tag);
+            if (!d.name().isEmpty()) {
+                builder.append(" " + d.name());
+            }
+            builder.append(" {");
+            builder.append("\n");
+            incr();
+        }
+        d.members().forEach(m -> m.accept(this, null));
+        if (!tag.isEmpty()) {
+            decr();
+            indent();
+            builder.append("};\n");
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitFunction(Declaration.Function d, Void ignored) {
+        indent();
+        var ftype = d.type();
+        var rtype = ftype.returnType();
+        builder.append(nameAndType(rtype, ""));
+        builder.append(" ");
+        builder.append(d.name());
+        builder.append("(");
+        builder.append(
+            d.parameters().
+                stream().
+                map(p -> nameAndType(p.type(), p.name())).
+                collect(Collectors.joining(", "))
+        );
+        if (d.type().varargs()) {
+            builder.append(",...");
+        }
+        builder.append(");\n");
+        return null;
+    }
+
+    @Override
+    public Void visitVariable(Declaration.Variable d, Void ignored) {
+        indent();
+        builder.append(nameAndType(d.type(), d.name()));
+        builder.append(";\n");
+        return null;
+    }
+
+    @Override
+    public Void visitConstant(Declaration.Constant d, Void ignored) {
+        indent();
+        builder.append(nameAndType(d.type(), d.name()));
+        builder.append(" = ");
+        Object value = d.value();
+        if (value instanceof String str) {
+            builder.append("\"" + Utils.quote(str) + "\"");
+        } else {
+            builder.append(value);
+        }
+        builder.append(";\n");
+        return null;
+    }
+
+    @Override
+    public Void visitTypedef(Declaration.Typedef d, Void ignored) {
+        indent();
+        builder.append("typedef ");
+        builder.append(nameAndType(d.type(), d.name()));
+        builder.append(";\n");
+        return null;
+    }
+
+    // In few cases, C type signature 'embeds' name.
+    // Examples:
+    //     int a[3]; // 'a' in between int and []
+    //     int (*func)(int); // 'func' is inside paren after '*'
+    // TypeVisitor accepts name and includes it in the appropriate
+    // place as needed. If not included, boolean flag is set to false
+    // in the result.
+
+    private static String nameAndType(Type type, String name) {
```

Comment:
Maybe we should call this "varDecl" since this is morally to be used for variable declarations.

Reply:
It's also not super clear to me what happens when the type has a name, and the user also specified a name. Can that ever happen? Or, if this happens, both names should be the same (in which case we should assert that).